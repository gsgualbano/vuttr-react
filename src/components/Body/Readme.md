#### Corpo da p√°gina, componente onde fica a lista de ferramentas.

Exemplo:

```js
<Body />
```
