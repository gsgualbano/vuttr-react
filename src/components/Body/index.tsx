import React from 'react';
import ReactLoading from 'react-loading';

import { connect } from 'react-redux';

import { Tool } from '~/store/ducks/tools/types';
import { ApplicationState } from '~/store';

import { 
  Container,
  Loading,
  ErrorMessage,
  EmptyToolsMessage,
} from './styles';

/**
 * Props que vem do redux.
 */
interface StateProps {
  /** array de tools */
  tools: Tool[];
  /** constante boolean que indica se ocorreu erro na requisição a api */
  loadError: boolean;
}

const Card = React.lazy<any>(() => import('../Card'));

/**
 * Corpo da página
 */
const Body = ({ tools, loadError }: StateProps): JSX.Element => {
  const renderLoadingIcon = (
    <Loading data-testid="load-spinner">
      <ReactLoading type="spin" color="#365DF0" height={70} width={70} />
    </Loading>
  );

  return (
    <Container>
      <React.Suspense fallback={renderLoadingIcon}>
        {tools.map(tool => (
          <Card key={tool.id} tool={tool} />
        ))}
      </React.Suspense>
      {tools.length === 0 && !loadError && <EmptyToolsMessage>No tools yet, to add a new tool press add!</EmptyToolsMessage>}
      {loadError && <ErrorMessage>Tools not found, try refreshing!</ErrorMessage>}
    </Container>
  );
};

const mapStateToProps = (state: ApplicationState) => ({
  tools: state.tools.data,
  loadError: state.tools.loadError,
});

export default connect(mapStateToProps)(Body);
