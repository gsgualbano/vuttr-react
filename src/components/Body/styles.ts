import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
`;

export const Loading = styled.div`
  margin: 0 auto;
`;

export const ErrorMessage = styled.p`
  margin-top: 20px;
  color: #f95e5a;
  text-align: center;
`;

export const EmptyToolsMessage = styled.p`
  margin-top: 20px;
  color: gray;
  text-align: center;
`;
