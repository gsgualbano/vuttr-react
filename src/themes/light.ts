export default {
  colors: {
    background: '#fcfcfd',
    textColor: '#170c3a',
    cardBackground: '#fcfcfd',
    cardBorderColor: '#ebeaed',
    modalBackground: '#fefefe',
    modalBorderColor: '#888',
    linkColor: 'rgb(0, 0, 238)',
    linkVisitedColor: 'rgb(85, 26, 139);',
    buttonOutlineBackground: '#ffffff',
    InputBackground: '#f5f4f6',
    InputBorderColor: '#ebeaed',
  },
};
