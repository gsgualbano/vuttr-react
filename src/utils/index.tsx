/* eslint-disable max-len */
import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import { render } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';
import { reducer } from '../store/ducks/tools';

interface Theme {
  colors: {
    background: string;
    textColor: string;
    cardBackground: string;
    cardBorderColor: string;
    modalBackground: string;
    modalBorderColor: string;
    linkColor: string;
    linkVisitedColor: string;
    buttonOutlineBackground: string;
    InputBackground: string;
    InputBorderColor: string;
  };
}

/** redux mock to render components that depends on the redux */
export const renderWithRedux = (
  component: JSX.Element,
  { initialState, store = createStore(reducer, initialState) }: any = {},
) => ({
  ...render(<Provider store={store}>{component}</Provider>),
  store,
});

/** render components thats depends on the theme provider */
export const renderWithTheme = (component: JSX.Element, theme: Theme) => render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);

/** render components thats depends on the redux and theme provider */
export const renderWithReduxAndTheme = (
  component: JSX.Element,
  theme: Theme,
  { initialState, store = createStore(reducer, initialState) }: any = {},
) => ({
  ...render(
    <ThemeProvider theme={theme}>
      <Provider store={store}>{component}</Provider>
    </ThemeProvider>,
  ),
  store,
});
