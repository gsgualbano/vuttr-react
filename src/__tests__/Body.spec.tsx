import React from 'react';
import { cleanup, wait } from '@testing-library/react';
import { renderWithReduxAndTheme } from '~/utils';

import Body from '../components/Body';

import 'jest-dom/extend-expect';
import 'jest-styled-components';

import light from '~/themes/light';

describe('<Body />', () => {
  afterEach(cleanup);

  it('renders correctly', () => {
    const { asFragment } = renderWithReduxAndTheme(<Body />, light, {
      initialState: {
        tools: {
          data: [],
          loadError: false,
        },
      },
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('render with load error', () => {
    const { getByText } = renderWithReduxAndTheme(<Body />, light, {
      initialState: {
        tools: {
          data: [],
          loadError: true,
        },
      },
    });
    expect(getByText('Tools not found, try refreshing!')).toBeInTheDocument();
  });

  it('render with load spinner', () => {
    const { getByTestId } = renderWithReduxAndTheme(<Body />, light, {
      initialState: {
        tools: {
          data: [
            {
              id: 1,
              title: 'Google',
              description: 'Search Website',
              link: 'https://google.com',
              tags: ['site', 'search', 'engine'],
            },
            {
              id: 2,
              title: 'Facebook',
              description: 'Social Network',
              link: 'https://facebook.com',
              tags: ['social', 'network', 'site'],
            },
          ],
          loadError: false,
        },
      },
    });

    expect(getByTestId('load-spinner')).toBeInTheDocument();
  });

  it('render with cards', async () => {
    const { queryAllByTestId, getByText, queryAllByText } = renderWithReduxAndTheme(
      <Body />,
      light,
      {
        initialState: {
          tools: {
            data: [
              {
                id: 1,
                title: 'Google',
                description: 'Search Website',
                link: 'https://google.com',
                tags: ['site', 'search', 'engine'],
              },
              {
                id: 2,
                title: 'Facebook',
                description: 'Social Network',
                link: 'https://facebook.com',
                tags: ['social', 'network', 'site'],
              },
            ],
            loadError: false,
          },
        },
      },
    );

    await wait();
    expect(queryAllByTestId('card-item')).toHaveLength(2);
    expect(getByText('Google')).toBeInTheDocument();
    expect(getByText('Facebook')).toBeInTheDocument();
    expect(getByText('Search Website')).toBeInTheDocument();
    expect(getByText('Social Network')).toBeInTheDocument();
    expect(getByText('#search')).toBeInTheDocument();
    expect(getByText('#engine')).toBeInTheDocument();
    expect(getByText('#social')).toBeInTheDocument();
    expect(getByText('#network')).toBeInTheDocument();
    expect(queryAllByText('#site')).toHaveLength(2);
  });
});
