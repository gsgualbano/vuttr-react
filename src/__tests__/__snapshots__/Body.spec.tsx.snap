// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Body /> renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  margin-top: 20px;
  color: gray;
  text-align: center;
}

<div
    class="c0"
  >
    <p
      class="c1"
    >
      No tools yet, to add a new tool press add!
    </p>
  </div>
</DocumentFragment>
`;
